src/core/lib/iomgr/ev_epoll_linux.c

```diff
@@ -1060,101 +1095,182 @@ static void fd_orphan(grpc_exec_ctx *exec_ctx, grpc_fd *fd,
+          is_done = false;
+          break;
+        }
+
+        case CLOSURE_NOT_READY: {
+          is_done = false;
+          break;
+        }
+
+        default: {
+          /* 'curr' is either a closure or the fd is already shutdown */
+          if ((curr & CLOSURE_SHUTDOWN) > 0) {
+            /* fd is already shutdown. Do nothing */
+          } else if (gpr_atm_rel_cas(state, curr, new_state)) {
+            grpc_closure_sched(
+                exec_ctx, (grpc_closure *)curr,
+                GRPC_ERROR_CREATE_REFERENCING("FD Shutdown", &shutdown_err, 1));
+          } else {
+            /* 'curr' was a closure but now changed to a different state. We
+                will have to retry */
+            is_done = false;
+          }
+          break;
+        }
+      }
+    }
   }
 }
 
-/* returns 1 if state becomes not ready */
-static int set_ready_locked(grpc_exec_ctx *exec_ctx, grpc_fd *fd,
-                            grpc_closure **st) {
-  if (*st == CLOSURE_READY) {
-    /* duplicate ready ==> ignore */
-    return 0;
-  } else if (*st == CLOSURE_NOT_READY) {
-    /* not ready, and not waiting ==> flag ready */
-    *st = CLOSURE_READY;
-    return 0;
-  } else {
-    /* waiting ==> queue closure */
-    grpc_closure_sched(exec_ctx, *st, fd_shutdown_error(fd));
-    *st = CLOSURE_NOT_READY;
-    return 1;
-  }
+static void set_ready(grpc_exec_ctx *exec_ctx, grpc_fd *fd, gpr_atm *state) {
+  /* Try the fast-path first (i.e expect the current value to be
+   * CLOSURE_NOT_READY */
+  if (!gpr_atm_acq_cas(state, CLOSURE_NOT_READY, CLOSURE_READY)) {
+    /* Fallback to slowpath */
+    gpr_atm curr = gpr_atm_acq_load(state);
+    switch (curr) {
+      case CLOSURE_READY: {
+        /* Already ready. We are done here */
+        break;
+      }
+
+      case CLOSURE_NOT_READY: {
+        /* The state was not CLOSURE_NOT_READY when we checked initially at the
+           beginning of this function but now it is CLOSURE_NOT_READY again.
+           This is only possible if the state transitioned out of
+           CLOSURE_NOT_READY to either CLOSURE_READY or <some closure> and then
+           back to CLOSURE_NOT_READY again (i.e after we entered this function,
+           the fd became "ready" and the necessary actions were already done).
+           So there is no need to make the state CLOSURE_READY now */
+        break;
+      }
+
+      default: {
+        /* 'curr' is either a closure or the fd is shutdown */
+        if ((curr & CLOSURE_SHUTDOWN) > 0) {
+          /* The fd is shutdown. Do nothing */
+        } else if (gpr_atm_rel_cas(state, curr, CLOSURE_NOT_READY)) {
+          grpc_closure_sched(exec_ctx, (grpc_closure *)curr, GRPC_ERROR_NONE);
+        }
+        /* else the state changed again (only possible by either a racing
+           set_ready or set_shutdown functions. In both these cases, the closure
+           would have been scheduled for execution. So we are done here */
+        break;
```

Comment:
Yeah. Retrying one more time would be harmless (as you mentioned, it would just be a spurious read.. but i couldn't think of any case either as to why we need to retry)